### DISCUSSION TITLE
Performance: Insert 1.000+ records

### DISCUSSION THREAD

**Discussion by MariuszKogut** [directus/directus#17545] [Archive]
Dear directus,

I'm integrating your headless cms into our IoT solution to manage some device data. For that I build a structure to manage some events (alarm, warning, information):

I have one relation to the device - it's many to one .
I have some translated text - it's build with the build in translations feature

Now I would like to import around 1.000 records to seed the data via graphql mutations.

I wrote an integration test to find out, how fast or slow Directus is. My results:

![image](https://user-images.githubusercontent.com/9637281/219267431-a191ac65-5d8c-4408-8ef4-0f12778ad1d0.png)

To add 1.000 events without translations, directus needs 4 secs.
But when I want to add 1.000 events with two translation-items, it becomes very slow (39 secs).

This is how my mutation looks like:

![image](https://user-images.githubusercontent.com/9637281/219267569-04a7a46c-a7b9-40e8-882c-3b6e766af353.png)

And this is how I'm passing the values:

![image](https://user-images.githubusercontent.com/9637281/219267632-909673db-471e-4b60-bada-ede11e6e7406.png)

I'm wondering if there is a way to improve the performance, when translations are involved.

Thanks,
Mariusz

**Comment by MariuszKogut** [comment]
@br41nslug Can you help me get in touch with the right person to discuss performance? 🙏

**Reply by br41nslug** [reply]
Are you running into the same performance issues when changing the insert method or is it specifically a graphql issue? For instance using REST or inserting data directly into the translation table instead of letting Directus figure out the relation.

**Reply by MariuszKogut** [reply]
Hey @br41nslug 

Thanks for your tipp. I was able to generate a C# client via Direcut's Open-API specification and access the REST API. Good job! 🌞

In my load test I'm creating 2.000 translations and 1.000 events. Each event is linked to 2 translations.

Here the performance result:

![image](https://user-images.githubusercontent.com/9637281/223362707-109d8ea0-e16a-4650-8815-792cbb4f09b9.png)

So it's even slower than using GraphQL in this case. 

Is this the desired performance throughout or is there maybe a bug related to translations? 

Thanks,
Mariusz

Used Directus version: 9.22.4 (hosted on docker)

Directus output:
```
2023-03-07 09:17:17 [08:17:17] POST /items/events_translations?meta= 200 33.1s
2023-03-07 09:18:01 [08:18:00] POST /items/events?meta= 200 43.2s
```

**Reply by MariuszKogut** [reply]
Hey @br41nslug,

updated to latest version ``9.23.1` without effects.

Bye,
Mariusz

**Comment by ctholho** [comment]
I looked into this after we moved database and directus to different nodes a few months ago. It took remarkably longer to send a `createMany` request with 60 objects as opposed to before.

Inspecting the database logs and looking at [/api/src/services/items.ts](https://github.com/directus/directus/blob/8ffb92ffa1690cc07b759e1f28fbc934e5caa8a5/api/src/services/items.ts#L324-L333)
it becomes obvious. Each object is sent individually to the db – and if all is well, the transaction is committed.

While it sounds like an obvious idea to send only one big request to the db, I get that it's not that easy, because you also want to update the revisions and relations and must guarantee that they have the correct id across all sql dialects with knex. I guess most sql implementations don't guarantee order when they return on `insert`.

Is this at all possible to implement without resorting to dialect specifics?

**Reply by rijkvanzanten** [reply]
> While it sounds like an obvious idea to send only one big request to the db, I get that it's not that easy, because you also want to update the revisions and relations and must guarantee that they have the correct id across all sql dialects with knex. I guess most sql implementations don't guarantee order when they return on insert.

Very true!

> Is this at all possible to implement without resorting to dialect specifics?

From my own experience, I'm afraid the answer is no.. At present time, that would mean adding a _lot_ of "if database == x, do y" inside of the API code, which is very hard to maintain over time. We're currently working on a new data-layer abstraction that sits in between though that will allow the API services logic to remain that single workflow, but then have that layer apply dialect specific additions / optimizations. That should then allow us to make the happy path as optimized as possible for each dialect, and only apply some of those hackier workarounds (like having to ensure that the insertion order is correct) for only some of the vendors 🙂
